File: qbopt/analysis/cellmap.py
"""Facts keyed by memory cell, indexed by the object each cell lies in."""

import bisect
import itertools
from collections.abc import Callable, Hashable, Iterable


class _Spans:
    """One bucket's cells by the displacement they start at.

    (low, high, cell), sorted by low once `sort` has run since the last
    out-of-order `add`. `widest` is the widest cell ever held, a bound on
    how far below a write a cell meeting it can start; it does not shrink
    when that cell goes.
    """

    __slots__ = ("cells", "sorted", "widest")

    def __init__(self) -> None:
        self.cells: list[tuple[int, int, Hashable]] = []
        self.sorted = True
        self.widest = 0

    def add(self, key, low: int, high: int) -> None:
        self.sorted = self.sorted and (not self.cells or self.cells[-1][0] <= low)
        self.cells.append((low, high, key))
        self.widest = max(self.widest, high - low)

    def sort(self) -> None:
        if not self.sorted:
            self.cells.sort(key=_low)
            self.sorted = True

    def remove(self, key, low: int) -> None:
        at = bisect.bisect_left(self.cells, low, key=_low) if self.sorted else 0
        while not (self.cells[at][0] == low and self.cells[at][2] == key):
            at += 1
        del self.cells[at]

    def meeting(self, low: int, high: int) -> list:
        """The cells whose bytes meet [low, high); `sort` has run."""
        found = []
        for start, end, key in itertools.islice(
            self.cells, bisect.bisect_left(self.cells, low - self.widest + 1, key=_low), None
        ):
            if start >= high:
                break
            if low < end:
                found.append(key)
        return found

    def copy(self) -> "_Spans":
        new = _Spans.__new__(_Spans)
        new.cells = list(self.cells)
        new.sorted = self.sorted
        new.widest = self.widest
        return new


def _low(cell: tuple) -> int:
    return cell[0]


class CellMap(dict):
    """A cell-to-fact dict that also buckets its cells by object.

    A write can only reach cells in objects it may alias, so `kill` tests
    those buckets and never the rest. Scanning every cell for every store
    ran deedlines' alias walk through 43M overlap tests in 90 s, all but
    1218 of them between different objects.

    `bucket_of` names a cell's bucket, a tuple; it must not change while
    the cell is held. `parts[i]` finds buckets by their i-th component, so a
    write can look its buckets up rather than test each one. `span_of`, if
    given, names the bytes [low, high) a cell covers, or None; a bucket's
    cells with bytes are indexed by them, for `kill`'s `displaced`.
    """

    __slots__ = ("bucket_of", "span_of", "buckets", "parts", "spans")

    def __init__(
        self,
        bucket_of: Callable[[Hashable], Hashable],
        items: Iterable = (),
        span_of: Callable[[Hashable], tuple[int, int] | None] | None = None,
    ) -> None:
        super().__init__()
        self.bucket_of = bucket_of
        self.span_of = span_of
        self.buckets: dict[tuple, set] = {}
        self.parts: list[dict[Hashable, set]] = []
        self.spans: dict[tuple, _Spans] = {}
        self.update(items)

    def __setitem__(self, key, value) -> None:
        if key not in self:
            bucket = self.bucket_of(key)
            keys = self.buckets.get(bucket)
            if keys is None:
                keys = self.buckets[bucket] = set()
                for i, part in enumerate(bucket):
                    if i == len(self.parts):
                        self.parts.append({})
                    self.parts[i].setdefault(part, set()).add(bucket)
            keys.add(key)
            if self.span_of is not None and (span := self.span_of(key)) is not None:
                spans = self.spans.get(bucket)
                if spans is None:
                    spans = self.spans[bucket] = _Spans()
                spans.add(key, *span)
        super().__setitem__(key, value)

    def __delitem__(self, key) -> None:
        super().__delitem__(key)
        bucket = self.bucket_of(key)
        keys = self.buckets[bucket]
        keys.discard(key)
        if self.span_of is not None and (span := self.span_of(key)) is not None:
            self.spans[bucket].remove(key, span[0])
        if not keys:
            del self.buckets[bucket]
            self.spans.pop(bucket, None)
            for i, part in enumerate(bucket):
                held = self.parts[i][part]
                held.discard(bucket)
                if not held:
                    del self.parts[i][part]

    def update(self, items=(), **named) -> None:
        pairs = items.items() if isinstance(items, dict) else items
        for key, value in pairs:
            self[key] = value
        for key, value in named.items():
            self[key] = value

    def copy(self) -> "CellMap":
        new = CellMap.__new__(CellMap)
        dict.update(new, self)
        new.bucket_of = self.bucket_of
        new.span_of = self.span_of
        new.buckets = {bucket: set(keys) for bucket, keys in self.buckets.items()}
        new.parts = [{part: set(held) for part, held in parts.items()} for parts in self.parts]
        new.spans = {bucket: spans.copy() for bucket, spans in self.spans.items()}
        return new

    def kill(
        self,
        reached: Iterable[Hashable] | None,
        overlaps: Callable[[Hashable], bool],
        displaced: tuple[set, int, int] | None = None,
    ) -> None:
        """Delete every cell a write reaches.

        `reached` are the buckets the write may touch (every bucket when
        None); `overlaps` is the exact test, asked only of their cells.
        `displaced` is (buckets, low, high): a cell in those buckets can
        only be reached if its bytes meet [low, high), so only those are asked.
        """
        buckets = self.buckets.keys() if reached is None else reached
        doomed = [key for bucket in buckets for key in self._asked(bucket, displaced) if overlaps(key)]
        for key in doomed:
            del self[key]

    def _asked(self, bucket, displaced: tuple[set, int, int] | None) -> Iterable:
        if displaced is not None and bucket in displaced[0] and (spans := self.spans.get(bucket)) is not None:
            spans.sort()
            return spans.meeting(displaced[1], displaced[2])
        return self.buckets.get(bucket, ())

    def setdefault(self, key, default=None):
        if key not in self:
            self[key] = default
        return self[key]

    def pop(self, key, *default):
        if key in self:
            value = self[key]
            del self[key]
            return value
        return dict.pop(self, key, *default)

    def popitem(self):
        raise TypeError("CellMap does not support popitem")

    def clear(self) -> None:
        super().clear()
        self.buckets.clear()
        self.parts.clear()
        self.spans.clear()

    def __ior__(self, other):
        self.update(other)
        return self

File: src/analysis/cellmap.rs
//! Facts keyed by memory cell, indexed by the object each cell lies in.
//!
//! Port of `qbopt/analysis/cellmap.py`. Python passes `bucket_of` to the
//! map; here the caller hands each new cell's bucket to `insert`, since a
//! bucket may come from a cache the caller also mutates.

use std::hash::Hash;
use std::ops::Deref;

use crate::support::hash::{HashMap, HashSet, IndexMap, IndexSet};

/// A cell's bucket, and how it indexes itself: Python's `parts[i]`, the
/// buckets held by their i-th component, so a write can look its buckets
/// up rather than test each one.
pub(crate) trait Bucket: Clone + Eq + Hash {
    type Parts: Default + Clone;
    /// Index a bucket that just gained its first cell.
    fn held(&self, parts: &mut Self::Parts);
    /// Unindex a bucket that just lost its last cell.
    fn released(&self, parts: &mut Self::Parts);
}

/// A cell-to-fact map that also buckets its cells by object.
///
/// A write can only reach cells in objects it may alias, so `kill` tests
/// those buckets and never the rest. A cell's bucket must not change while
/// the cell is held.
#[derive(Clone)]
pub(crate) struct CellMap<K, V, B: Bucket> {
    items: IndexMap<K, V>,
    pub buckets: HashMap<B, IndexSet<K>>,
    pub parts: B::Parts,
}

impl<K: Clone + Eq + Hash, V, B: Bucket> CellMap<K, V, B> {
    pub(crate) fn new(items: IndexMap<K, V>, mut bucket_of: impl FnMut(&K) -> B) -> Self {
        let mut map = Self { items: IndexMap::default(), buckets: HashMap::default(), parts: B::Parts::default() };
        for (key, value) in items {
            map.insert(key, value, &mut bucket_of);
        }
        map
    }

    /// Python `__setitem__`: `bucket_of` is asked only of a new cell.
    pub(crate) fn insert(&mut self, key: K, value: V, bucket_of: impl FnOnce(&K) -> B) {
        if !self.items.contains_key(&key) {
            let bucket = bucket_of(&key);
            let keys = self.buckets.entry(bucket.clone()).or_insert_with(|| {
                bucket.held(&mut self.parts);
                IndexSet::default()
            });
            keys.insert(key.clone());
        }
        self.items.insert(key, value);
    }

    /// Delete every cell a write reaches.
    ///
    /// `reached` are the buckets the write may touch (every bucket when
    /// None); `overlaps` is the exact test, asked only of their cells.
    pub(crate) fn kill(&mut self, reached: Option<HashSet<B>>, mut overlaps: impl FnMut(&K) -> bool) {
        let mut doomed = Vec::new();
        let mut test = |bucket: &B, keys: &IndexSet<K>| {
            doomed.extend(keys.iter().filter(|key| overlaps(key)).map(|key| (bucket.clone(), key.clone())));
        };
        match &reached {
            None => self.buckets.iter().for_each(|(bucket, keys)| test(bucket, keys)),
            Some(reached) => reached
                .iter()
                .filter_map(|bucket| self.buckets.get(bucket).map(|keys| (bucket, keys)))
                .for_each(|(bucket, keys)| test(bucket, keys)),
        }
        #[cfg(test)]
        if let Some(reached) = &reached {
            if VERIFYING.with(std::cell::Cell::get) {
                for (bucket, keys) in &self.buckets {
                    if !reached.contains(bucket) {
                        VERIFIED.with(|checked| checked.set(checked.get() + keys.len()));
                        assert!(!keys.iter().any(&mut overlaps), "a skipped bucket holds a cell the write reaches");
                    }
                }
            }
        }
        if doomed.is_empty() {
            return;
        }
        for (bucket, key) in &doomed {
            let keys = self.buckets.get_mut(bucket).expect("a doomed cell's bucket is held");
            keys.swap_remove(key);
            if keys.is_empty() {
                self.buckets.remove(bucket);
                bucket.released(&mut self.parts);
            }
        }
        // Python's `del` keeps the rest in order; so does one retain.
        let gone = doomed.into_iter().map(|(_, key)| key).collect::<HashSet<_>>();
        self.items.retain(|key, _| !gone.contains(key));
    }

    pub(crate) fn into_items(self) -> IndexMap<K, V> {
        self.items
    }
}

#[cfg(test)]
thread_local! {
    /// While set, `kill` asks the exact test of every cell it skipped and
    /// counts them in `VERIFIED`: the index may only skip work.
    pub(crate) static VERIFYING: std::cell::Cell<bool> = const { std::cell::Cell::new(false) };
    pub(crate) static VERIFIED: std::cell::Cell<usize> = const { std::cell::Cell::new(0) };
}

impl<K, V, B: Bucket> Deref for CellMap<K, V, B> {
    type Target = IndexMap<K, V>;

    fn deref(&self) -> &IndexMap<K, V> {
        &self.items
    }
}

#[cfg(test)]
#[path = "cellmap_tests.rs"]
mod tests;
